### PR TITLE
Update utils.py and Create test_utils_predict.py

### LINKED ISSUE #565: Should change np.hstack to np.vstack.
https://github.com/tensorlayer/tensorlayer/blob/4db0b9f52f629742967eb05e3ddb1a23a5ca666f/tensorlayer/utils.py#L323

### PR DESCRIPTION
**Pull request by 2wins**
### Checklist
- [x] I've tested that my changes are compatible with the latest version of Tensorflow.
- [x] I've read the [Contribution Guidelines](https://github.com/tensorlayer/tensorlayer/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context

Fix #565
In #288, the error has not been handled properly.

### Description

`np.hstack` makes data stacked along `axis=1` while `np.vstack` along `axis=0`.

### CODE REVIEW

**Comment by zsdonghao** [comment]
@2wins thanks, I have one question. If the number of example can not be divided by the batch size, it is still work fine?

**Comment by 2wins** [comment]
@zsdonghao Partially yes. Please check [my comment](https://github.com/tensorlayer/tensorlayer/issues/288#issuecomment-388623558) in #288.

**Comment by DEKHTIARJonathan** [comment]
@2wins @zsdonghao why not trying to write a small unittests for all the cases you could think of?
If it is something tricky it would definitely be a good thing to have it tested.

**Comment by zsdonghao** [comment]
@2wins nice! it would be great to add this test scripts into test folder ~

**Comment by 2wins** [comment]
@DEKHTIARJonathan I see. Sorry to bother you.
@zsdonghao BTW, how can I join the Slack? Every invitation has not been sent.

**Comment by DEKHTIARJonathan** [comment]
There is a huge button in the readme. There is no way u can miss it.

**Comment by DEKHTIARJonathan** [comment]
BTW. As you can see the test is still failing. So there is an error either in your modifications or your test 😉.
That's exactly why I asked you to implement a test 😊

**Comment by 2wins** [comment]
@DEKHTIARJonathan OK, I'll check it.

**Comment by DEKHTIARJonathan** [comment]
@2wins 
image

File: CHANGELOG.md [hunk 1]
```diff
@@ -72,9 +72,13 @@ To release a new version, please update the changelog as followed:
 - Repository:
   - Danger CI has been added to enforce the update of the changelog (by @lgarithm and @DEKHTIARJonathan)
 - Tutorials:
-  - `tutorial_tfslim` added by (@2wins).
+  - `tutorial_tfslim` added (by @2wins).
+- Tests:
+  - `test_utils_predict.py` added (by @2wins).
```
Comment (DEKHTIARJonathan) [inline CHANGELOG.md]: Please mention why the test has been added: to insure that issue #??? is solved.

**Comment by 2wins** [comment]
@DEKHTIARJonathan `case 2` and `case 3` make an error. So I commented them out.

**Comment by DEKHTIARJonathan** [comment]
@2wins That's not how it is supposed to work.
The objective of a test is to highlight the errors and bugs. If something is not working, you should let the tests as it is supposed to be. And fix the codebase.

**Comment by 2wins** [comment]
@DEKHTIARJonathan I tried joining Slack a few times, but the mail does not come.

File: CHANGELOG.md (hunk 1 above)
Comment (DEKHTIARJonathan) [reply CHANGELOG.md]: You can do the following:

```
- `test_utils_predict.py` added to reproduce and fix issue #288 (by @2wins).
```

**Comment by 2wins** [comment]
@DEKHTIARJonathan I see. Many thanks.

File: CHANGELOG.md [hunk 2]
```diff
@@ -72,9 +72,13 @@ To release a new version, please update the changelog as followed:
...
   - Danger CI has been added to enforce the update of the changelog (by @lgarithm and @DEKHTIARJonathan)
 - Tutorials:
-  - `tutorial_tfslim` added by (@2wins).
+  - `tutorial_tfslim` added (by @2wins).
+- Tests:
+  - `test_utils_predict.py` added to reproduce and fix issue #288 (by @2wins).
 
 ### Changed
+- Util:
+  - `predict` function modified to fix #565 (by @2wins).
```
Comment (DEKHTIARJonathan) [inline CHANGELOG.md]: If it is a bug fix => It should go to the section. [Fixed](https://github.com/2wins/tensorlayer/blob/5ccc7a912331cc5061a7722af0f448049d0a941b/CHANGELOG.md#fixed)

I would do something like this:
```
- Issue #288 related to `tl.utils.predict` Fixed - predict np.vstack problem when batch_size if no factor of test dataset size
```

File: CHANGELOG.md [hunk 3]
```diff
@@ -77,14 +77,14 @@ To release a new version, please update the changelog as followed:
...
-- Util:
-  - `predict` function modified to fix #565 (by @2wins).
 
 ### Deprecated
 
 ### Removed
 
 ### Fixed
+- Util:
+  - `predict` function modified to fix #565 (by @2wins).
```
Comment (DEKHTIARJonathan) [inline CHANGELOG.md]: You did not follow my comment:
````
- Issue #288 related to `tl.utils.predict` Fixed - predict np.vstack problem when batch_size if no factor of test dataset size
```
This is a lot more precise.

Comment (2wins) [reply CHANGELOG.md]: @DEKHTIARJonathan Sorry, I'll follow you.

Comment (DEKHTIARJonathan) [reply CHANGELOG.md]: I think you can just copy+paste my example, it is way enough. We do not have hundreds of bug fixes per release (hopefully).

Comment (2wins) [reply CHANGELOG.md]: I read your mention and I think it should be modified as follows:
- Issue #565 related to `tl.utils.predict` fixed
  - predict `np.hstack` problem in which the results for multiple batches are stacked along `dim=1`

Comment (DEKHTIARJonathan) [reply CHANGELOG.md]: Why would you create a sublist for only one point ?
I would absolutely agree with if there are many points that you would like to mention is it the case?

I mean something like this:

``` 
- Issue #565 related to `tl.utils.predict` fixed
  - predict `np.hstack` problem in which the results for multiple batches are stacked along `dim=1`
  - reason 2
  - ...
```

Comment (2wins) [reply CHANGELOG.md]: OK, it should be 
- Issue #565 related to `tl.utils.predict` fixed - predict `np.hstack` problem in which the results for multiple batches are stacked along `dim=1`

**Comment by DEKHTIARJonathan** [comment]
Okay now we can work on the core of the PR:
1. You should fix the test file to use `unittest` and be coherent with the rest of the project
2. We will see how we can fix `tl.utils.predict` in order to make sure that the test pass.

**Comment by 2wins** [comment]
@DEKHTIARJonathan @zsdonghao Please solve the Slack problem. I'd like to ask you guys some questions in Slack.

**Comment by DEKHTIARJonathan** [comment]
I have no control over this.

@zsdonghao why is Slack only under invitation? Can't we open it fully?

**Comment by zsdonghao** [comment]
@2wins @DEKHTIARJonathan the invitation link doesn't expire, can you try again ?

**Comment by 2wins** [comment]
@zsdonghao It still doesn't work. :cry:

**Comment by zsdonghao** [comment]
@2wins can you try this?

https://join.slack.com/t/tensorlayer/shared_invite/enQtMjUyMjczMzU2Njg4LWI0MWU0MDFkOWY2YjQ4YjVhMzI5M2VlZmE4YTNhNGY1NjZhMzUwMmQ2MTc0YWRjMjQzMjdjMTg2MWQ2ZWJhYzc

**Comment by 2wins** [comment]
@zsdonghao It does not work as well.

**Comment by zsdonghao** [comment]
@2wins what is your slack account, I add you first.

**Comment by 2wins** [comment]
@zsdonghao my email is  twinkyu@live.com.

**Comment by zsdonghao** [comment]
@2wins invite send

**Comment by 2wins** [comment]
@zsdonghao I haven't had any email from Slack. 😭

**Comment by 2wins** [comment]
@DEKHTIARJonathan @zsdonghao Please check the Travis-CI error and give me a advice.

File: tests/test_utils_predict.py [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+
+class Util_Predict_Test(CustomTestCase):
+
+    @classmethod
+    def setUpClass(cls):
+        cls.x1 = tf.placeholder(tf.float32, [None, 5, 5, 3])
+        cls.x2 = tf.placeholder(tf.float32, [8, 5, 5, 3])
+        cls.X1 = np.ones([127, 5, 5, 3])
+        cls.X2 = np.ones([7, 5, 5, 3])
```
[the remark this answers is not in the document]
Comment (DEKHTIARJonathan) [reply tests/test_utils_predict.py]: This is the error, I was correct: 

https://travis-ci.org/tensorlayer/tensorlayer/jobs/378752347#L1005

image

[the remark this answers is not in the document]
Comment (2wins) [reply tests/test_utils_predict.py]: Sometimes we set the shape of `tf.placeholder` explicitly using a specific batch size.
However, it becomes a problem when the number of samples is not a factor of the batch size.

[the remark this answers is not in the document]
Comment (2wins) [reply tests/test_utils_predict.py]:
```python
cls.x2 = tf.placeholder(tf.float32, [8, 5, 5, 3])
cls.X2 = np.ones([7, 5, 5, 3])
```
```python
# Actually, the same with the above case
cls.x2 = tf.placeholder(tf.float32, [8, 5, 5, 3])
cls.X2 = np.ones([135, 5, 5, 3])
```

**Comment by DEKHTIARJonathan** [comment]
Are you sure to understand how TF placeholders works?
To my best knowledge, it is simply impossible to input [7, 5, 5, 3] into a plh of shape [8, 5, 5, 3]
@zsdonghao do you confirm?

**Comment by 2wins** [comment]
@DEKHTIARJonathan That's the problem of `tl.utils.predict`. Of course, we cannot feed data into `tf.placeholder` with non-matched size. I think there are two options for `tl.utils.predict`.

1. not allow the function to take data with non-matched size.
2. using dummy data to fit input size.

**Comment by DEKHTIARJonathan** [comment]
I'm not sure to understand. Why do we want to fix this issue ?
It is not a correct practice to input non-matching data size into a placeholders. If the PLH size varies, you have to use `None` as a batch_size dimension.

Is there any case where the user do not control the placeholder and need to be able custom batch size ?

**Comment by 2wins** [comment]
@DEKHTIARJonathan I agree with you. However, some beginners could encounter this problem as seen in issue #288 and may think it as tensorlayer's problem. What do you think of it?

**Comment by DEKHTIARJonathan** [comment]
@2wins I believe that everyone can make mistakes. Maybe we can raise an explicit Exception "Batch_Size Mismatched bla bla bla". However, we should **not** in any case torture the library to make it user-error proof.

**Comment by 2wins** [comment]
@DEKHTIARJonathan You're right. I have the same opinion about that. I think the explicit exception is the best solution here. Thanks.

**Comment by 2wins** [comment]
@DEKHTIARJonathan Thanks. I'll apply it.

**Comment by DEKHTIARJonathan** [comment]
@2wins Please resolve the conflicts by copy-pasting the following, I can not do it on your fork.
```
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/)
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

<!--

============== Guiding Principles ==============

* Changelogs are for humans, not machines.
* There should be an entry for every single version.
* The same types of changes should be grouped.
* Versions and sections should be linkable.
* The latest version comes first.
* The release date of each version is displayed.
* Mention whether you follow Semantic Versioning.

============== Types of changes (keep the order) ==============

* `Added` for new features.
* `Changed` for changes in existing functionality.
* `Deprecated` for soon-to-be removed features.
* `Removed` for now removed features.
* `Fixed` for any bug fixes.
* `Security` in case of vulnerabilities.
* `Dependencies Update` in case of vulnerabilities.
* `Contributors` to thank the contributors that worked on this PR.

============== How To Update The Changelog for a New Release ==============

** Always Keep The Unreleased On Top **

To release a new version, please update the changelog as followed:
1. Rename the `Unreleased` Section to the Section Number
2. Recreate an `Unreleased` Section on top
3. Update the links at the very bottom

======================= START: TEMPLATE TO KEEP IN CASE OF NEED ===================

** DO NOT MODIFY THIS SECTION ! **

## [Unreleased]

### Added

### Changed

### Deprecated

### Removed

### Fixed

### Security

### Dependencies Update

### Contributors

** DO NOT MODIFY THIS SECTION ! **

======================= END: TEMPLATE TO KEEP IN CASE OF NEED ===================

-->

<!-- YOU CAN EDIT FROM HERE -->

## [Unreleased]

### Added
- Tutorials:
  - `tutorial_tfslim` has been introduced to show how to use `SlimNetsLayer` (by @2wins in #560).
- Test:
  - `Layer_DeformableConvolution_Test` added to reproduce issue #572 with deformable convolution (by @DEKHTIARJonathan in #573)  
  - `test_utils_predict.py` added to reproduce and fix issue #288 (by @2wins in #566)
- CI Tool:
  - Danger CI has been added to enforce the update of the changelog (by @lgarithm and @DEKHTIARJonathan in #563)
  - https://github.com/apps/stale/ added to clean stale issues (by @DEKHTIARJonathan in #573)
>>>>>>> master

### Changed
- Tensorflow CPU & GPU dependencies moved to separated requirement files in order to allow PyUP.io to parse them (by @DEKHTIARJonathan in #573)

### Deprecated

### Removed

### Fixed
- Issue #498 - Deprecation Warning Fix in `tl.layers.RNNLayer` with `inspect` (by @DEKHTIARJonathan in #574)
- Issue #498 - Deprecation Warning Fix in `tl.files` with truth value of an empty array is ambiguous (by @DEKHTIARJonathan in #575)
- Issue #572 with deformable convolution fixed (by @DEKHTIARJonathan in #573)
- Issue #565 related to `tl.utils.predict` fixed - `np.hstack` problem in which the results for multiple batches are stacked along `dim=1` (by @2wins in #566)

### Security

### Dependencies Update

### Contributors
@lgarithm @DEKHTIARJonathan @2wins


## [1.8.5] - 2018-05-09

### Added
- Github Templates added (by @DEKHTIARJonathan)
  - New issues Template
  - New PR Template
- Travis Deploy Automation on new Tag (by @DEKHTIARJonathan)
  - Deploy to PyPI and create a new version.
  - Deploy to Github Releases and upload the wheel files
- PyUP.io has been added to ensure we are compatible with the latest libraries (by @DEKHTIARJonathan)
- `deconv2d` now handling dilation_rate (by @zsdonghao)
- Documentation unittest added (by @DEKHTIARJonathan)
- `test_layers_core` has been added to ensure that `LayersConfig` is abstract.

### Changed
- All Tests Refactored - Now using unittests and runned with PyTest (by @DEKHTIARJonathan)
- Documentation updated (by @zsdonghao)
- Package Setup Refactored (by @DEKHTIARJonathan)
- Dataset Downlaod now using library progressbar2 (by @DEKHTIARJonathan)
- `deconv2d` function transformed into Class (by @zsdonghao)
- `conv1d` function transformed into Class (by @zsdonghao)
- super resolution functions transformed into Class (by @zsdonghao)
- YAPF coding style improved and enforced (by @DEKHTIARJonathan)

### Fixed
- Backward Compatibility Restored with deprecation warnings (by @DEKHTIARJonathan)
- Tensorflow Deprecation Fix (Issue #498):
  - AverageEmbeddingInputlayer (by @zsdonghao)
  - load_mpii_pose_dataset (by @zsdonghao)
- maxPool2D initializer issue #551 (by @zsdonghao)
- `LayersConfig` class has been enforced as abstract
- Pooling Layer Issue #557 fixed (by @zsdonghao)

### Dependencies Update
- scipy>=1.0,<1.1 => scipy>=1.1,<1.2

### Contributors
@zsdonghao @luomai @DEKHTIARJonathan

[Unreleased]: https://github.com/tensorlayer/tensorlayer/compare/1.8.5...master
[1.8.5]: https://github.com/tensorlayer/tensorlayer/compare/1.8.4...1.8.5

```

**Comment by DEKHTIARJonathan** [comment]
@2wins you have to use the resolve conflicts function 

image

Please copy+paste exactly the code I gave you by replacing everything ;)

**Comment by 2wins** [comment]
@DEKHTIARJonathan I tried it first but `Mark as resolved` button was inactive even if there was no conflict. Now I've done it. Many Thanks.

File: CHANGELOG.md [hunk 5]
```diff
@@ -72,6 +72,8 @@ To release a new version, please update the changelog as followed:
 - Tutorials:
   - `tutorial_tfslim` has been introduced to show how to use `SlimNetsLayer` (by @2wins in #560).
 - Test:
+  - `Layer_DeformableConvolution_Test` added to reproduce issue #572 with deformable convolution (by @DEKHTIARJonathan in #573)
```
Comment (DEKHTIARJonathan) [inline CHANGELOG.md]: This line is a duplicate, please remove it

**Comment by lgarithm** [comment]
I don't use **numpy** much, but it looks OK from the [document](https://docs.scipy.org/doc/numpy/reference/generated/numpy.hstack.html).
And it actually fixed the referred issue in comment.

LGTM

(Not that the layout of the **result** variable is changed.)


@zsdonghao I'll leave it to you to merge.

**Comment by 2wins** [comment]
@DEKHTIARJonathan FYI,
- `concatenate`: Join a sequence of arrays along an existing axis.
- `hstack`: Stack arrays in sequence horizontally (column wise)
- `vstack`: Stack arrays in sequence vertically (row wise)

`hstack` and `vstack` are implemented using `concatenate`.

**Comment by 2wins** [comment]
@zsdonghao Please confirm this PR. Thanks.

**Pull request merged**